### PR TITLE
Move functions gathering AMR tree to separate location

### PR DESCRIPTION
**Pull request by KunkelAlexander**
# Move functions gathering AMR tree to separate location

This pull request moves the functions for synchronising the patch counts and the load balancing ID's between different ranks as well as the functions for calculating the GIDs of father, son and sibling patches the new source file LB_GatherTree.cpp.
The corresponding structures are defined in patch.h, but can be moved to a more appropriate header file if necessary.
The pull request further modifies the code for storing the HDF5 file and the YT inline analysis in order to reduce code redundancy.
In addition, it provides a function that generates a flat tree structure LB_GlobalPatch global_tree[NPatchAllLv] where
```

struct LB_GlobalPatch
{
   int    corner[3];
   long    sibling[26];
   long    father;
   long    son;
   long   LB_Idx;
   int    level; 
#  ifdef PARTICLE
   int    NPar;
#  endif 
}; // struct LB_GlobalPatch
```

stores the corresponding GIDs of sibling, father and son patches and a patch with a given GID can be retrieved as global_tree[GID]. The tree can either be distributed to all ranks (Allgather) or to the root rank (Gather).
The code was tested in HYDRO and ELBDM in SERIAL mode and with LOAD_BALANCING and OpenMP.
Both the HDF5 files and the YT inline analysis seem to work correctly.

### CODE REVIEW

File: include/Patch.h
Comment (hyschive) [inline include/Patch.h:1154]: Missing `;`

File: src/Output/Output_DumpData_Total_HDF5.cpp
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:1332]: `const int SonGID`

File: src/Output/Output_DumpData_Total_HDF5.cpp
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:1354]: `const int SibGID`

**Review by hyschive — CHANGES_REQUESTED** [review]
@KunkelAlexander Looks great and very useful! I've left some minor comments.
- [ ] Check all inline comments.
- [ ] Remove all trailing spaces.
- [ ] I would suggest moving all new structures from `Patch.h` to a new header and including it in `GAMER.h`.
- [ ] Make sure that the code can be compiled successfully with `PARTICLE` and `GAMER_DEBUG`.

**Comment by cindytsai** [comment]
@hyschive @koarakawaii 
`SERIAL` option in gamer probably won't work. `libyt` currently only compiles with MPI. It only supports `mpirun -np 1` now (fake serial ...).
This is also something I will do in the future.

**Comment by hyschive** [comment]
> @hyschive @koarakawaii `SERIAL` option in gamer probably won't work. `libyt` currently only compiles with MPI. It only supports `mpirun -np 1` now (fake serial ...). This is also something I will do in the future.

@cindytsai Thanks for the comment! I've created #172 to track it.

**Comment by KunkelAlexander** [comment]
@cindytsai May I ask which `SERIAL` mode you mean? Because I have been testing my modifications to the `YT` routines in `Gamer` in `SERIAL` mode and with OpenMPI and so far all the figures look fine. Maybe I am using an old version of `libyt`?

**Comment by cindytsai** [comment]
@KunkelAlexander What I mean by **SERIAL** mode is compiled GAMER through `g++` when **SUPPORT_LIBYT** is switched on.

**Comment by KunkelAlexander** [comment]
@cindytsai Thank you for your reply. But that is strange since that is exactly what I am doing and it seems to work just fine. I do not get any compile-time errors and the plots generated by libyt look fine too. But maybe I am missing something.

File: include/Patch.h
Comment (KunkelAlexander) [reply include/Patch.h]: Yes, you are right about this. I was not consistent. I personally do not have a preference here and the only case where it would cause a problem is if we had a simulation with more than 2^32 - 1 = 2,147,483,647. I doubt that is feasible. So, I guess that an integer should be safe here. Should I also change the corresponding long datatypes in YT?

Comment (KunkelAlexander) [reply include/Patch.h]: I have changed the datatype to int in my code now. With this all your comments should have been fixed. I have tested the code in SERIAL mode (HDF5 and libYT) and will now run it with MPI.

Comment (KunkelAlexander) [reply include/Patch.h]: @cindytsai Did you test it and find out you had to use the long datatype for GIDs?

Comment (hyschive) [reply include/Patch.h]: @KunkelAlexander `int` is definitely OK for GAMER since each patch already contains at least `8^3` cells. `libyt` is more general-purpose so I'm not sure. But as long as this PR is still compatible with `libyt` it should be fine (i.e., make sure that we _copy_ GID instead of passing the entire GID-related _array_ to `libyt`.

Comment (KunkelAlexander) [reply include/Patch.h]: @hyschive Got it, then I will leave the `long` types in the `libyt`-related code untouched. I also realised that `libyt` expects function pointers to have `long` types as arguments.

**Comment by hyschive** [comment]
@KunkelAlexander Let me know when this PR is ready for review again.

**Comment by KunkelAlexander** [comment]
@hyschive I believe it is ready for review again.

**Comment by cindytsai** [comment]
@KunkelAlexander Can you please send me your `Makefile`, and the commit SHA of libyt you are using? :sweat_smile:

**Comment by KunkelAlexander** [comment]
@cindytsai I pulled my version of libyt from https://github.com/calab-ntu/libyt.git and the short SHA of the commit ID I am using is fd517674658b6812a42a9c999523484b3c81bc6c. It is from April 2022 :laughing: So, probably too old to be useful for debugging. I am not quite sure how I could install such an old version. Sorry for the confusion. I just assumed I had installed a halfway-recent version.

**Comment by KunkelAlexander** [comment]
And my makefile is really just the standard GAMER makefile with the HYDRO test in mode SERIAL with libyt support and the appropriate library paths.

**Comment by cindytsai** [comment]
Hi @KunkelAlexander , that version may be old, but not that old. I just try to compile GAMER using only `g++` with `libyt` compile using `mpic++`, and they worked :laughing: !! Thanks for pointing out.

---

Interestingly, I use `libyt` (compiled by `mpic++`) to build some small executable compiled by `g++` and `mpic++`.

```bash
mpic++ -I$INCLUDE_DIR example_libyt.cpp -L$LIBRARY_DIR -lyt
g++ -I$INCLUDE_DIR example_libyt.cpp -L$LIBRARY_DIR -lyt
```

- `mpic++` (success)
- `g++` (failed)
  ```c++
  #include <stdio.h>
  #include <mpi.h>   // comment out in g++
  #include "libyt.h"
  #include "libyt_interactive_mode.h"

  int main() {
      MPI_Init(&argc, &argv);  // comment out in g++

      yt_param_libyt param_libyt;
      param_libyt.verbose = YT_VERBOSE_INFO;
      param_libyt.script = "inline";
      param_libyt.check_data = false;

      if (yt_initialize(argc, argv, &param_libyt) != YT_SUCCESS) {
          printf("yt_initialize failed!\n");
          return 1;
      }

      if (yt_run_InteractiveMode("LIBYT_STOP") != YT_SUCCESS) {
          printf("yt_interactive_mode failed!\n");
          return 1;
      }

      MPI_Finalize();  // comment out in g++
      
      return 0;
  }
  ```

Error message in `g++`:
```
[YT_INFO   ] Initializing libyt ...
[YT_INFO   ]    verbose = 1
[YT_INFO   ]     script = inline
[YT_INFO   ]    counter = 0
[YT_INFO   ] check_data = false
*** The MPI_Comm_size() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[xps:38263] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

@hyschive Do you have any idea why is GAMER supporting compiling using `g++` :rofl: ?

**Comment by hyschive** [comment]
@cindytsai That does sound strange and I'm not sure why. Please keep me posted.

BTW, you may want to check [MPI_Initialized](https://www.open-mpi.org/doc/v4.1/man3/MPI_Initialized.3.php) in `libyt` before invoking any other MPI routines.

Let's move the discussions to gamer-project/gamer#172. I'll copy and paste some relevant information first.

(@KunkelAlexander)

File: src/LoadBalance/LB_GatherTree.cpp
Comment (hyschive) [inline src/LoadBalance/LB_GatherTree.cpp:680]: Replace `PID` with `GID`

File: src/LoadBalance/LB_GatherTree.cpp
Comment (hyschive) [inline src/LoadBalance/LB_GatherTree.cpp:670]: Add `LB_PatchCount pc;`

File: src/LoadBalance/LB_GatherTree.cpp
Comment (hyschive) [inline src/LoadBalance/LB_GatherTree.cpp:711]: Also, clarify that this function will set and return `pc` as well.

**Review by hyschive — COMMENTED** [review]
@KunkelAlexander Looks great! I have only a few very minor comments and would be happy to merge this PR right away.

**Comment by KunkelAlexander** [comment]
@hyschive Great. I am happy to hear that and thank you very much for your review. But maybe I should test it with the most recent version of libyt before we merge. 

@cindytsai I am glad you now have an idea as to why there may be problems with the SERIAL mode. :)

**Review by hyschive — APPROVED** [review]

**Comment by hyschive** [comment]
> But maybe I should test it with the most recent version of libyt before we merge.

@KunkelAlexander Good idea. I'll merge it once your test passes.

**Comment by koarakawaii** [comment]
@KunkelAlexander : Hi, I have some problem when testing this PR by the Plummer cloud(s) test problem. It seems that error related to `('gamer', 'density')` field cannot be found will occur when running the inline yt analysis, as shown below:
```
P000 yt : [ERROR    ] 2023-02-06 16:29:23,043 YTFieldNotFound: Could not find field ('gamer', 'density') in Fig000000000.
Did you mean:
        ('gas', 'density')
P000 yt : [ERROR    ] 2023-02-06 16:29:23,043 Error occurred on rank 0.
  File "<string>", line 1, in <module>
  File "./inline_script.py", line 29, in yt_inline_inputArg
    sz.set_zlim( ('gamer', fields), 4e-7, 5e-2 )
  File "/work1/koarakawaii/python3.7/python3.7/lib/python3.7/site-packages/yt/visualization/plot_container.py", line 136, in newfunc
    for f in self.data_source._determine_fields(field):
  File "/work1/koarakawaii/python3.7/python3.7/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1439, in _determine_fields
    finfo = self.ds._get_field_info(ftype, fname)
  File "/work1/koarakawaii/python3.7/python3.7/lib/python3.7/site-packages/yt/data_objects/static_output.py", line 834, in _get_field_info
    field_info, is_ambiguous = self._get_field_info_helper(ftype, fname)
  File "/work1/koarakawaii/python3.7/python3.7/lib/python3.7/site-packages/yt/data_objects/static_output.py", line 912, in _get_field_info_helper
    raise YTFieldNotFound(field=INPUT, ds=self)
```
but this error seems not happen when I tested the same test problem using the `GAMER` `main` branch. Is this related to modifications in `src/YT` or something else~?

File: src/YT/YT_Inline.cpp
Comment (KunkelAlexander) [inline src/YT/YT_Inline.cpp:265]: @koarakawaii There is this line of code that I changed because it gave errors with my version of libyt, but probably that was because my version was too old (?). I will change it back and push it again.

Comment (KunkelAlexander) [reply src/YT/YT_Inline.cpp:265]: @koarakawaii On a similar note, I also had to rename the field in the yt_inline.py plotting script. I hope that everything is consistent now. Could you please check again? I pushed the version where I changed the name of the field back to "Dens".  Thank you

Comment (koarakawaii) [reply src/YT/YT_Inline.cpp:265]: @KunkelAlexander: Thank you for the help! After changing this, the test can be run with this PR :grinning: !

Comment (KunkelAlexander) [reply src/YT/YT_Inline.cpp:265]: @koarakawaii  Thank you so much for your help :) I guess I am happy then. ;)

**Comment by koarakawaii** [comment]
@hyschive @KunkelAlexander : the figures produced by this PR and `GAMER` `main` branch are the same, so I think the PR passed the `libyt` test~

**Review by hyschive — APPROVED** [review]

**Pull request merged**
